### PR TITLE
refactor(menuable): convert to typescript

### PR DESCRIPTION
**Pull request by MajesticPotatoe**
## Description
Menuable - Last of the mixins to be converted. Looked lonely.

## Motivation and Context
Convert framework to TS

## How Has This Been Tested?
jest

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Improvement/refactoring (non-breaking change that doesn't add any feature but make things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and breaking changes).
- [x] My code follows the code style of this project.
- [ ] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

### CODE REVIEW

File: packages/vuetify/src/mixins/menuable.ts [hunk 1]
```diff
@@ -1,12 +1,16 @@
...
+// Utilities
 import { consoleError } from '../util/console'
+import mixins, { ExtractVue } from '../util/mixins'
+
+// Types
+import Vue, { VNode } from 'vue'
 
 /* eslint-disable object-property-newline */
-const dimensions = {
+const dimensions: any = {
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/mixins/menuable.ts]: not really `any`, it has precisely described structure, i think that the type should be imferred from the oject structure, didn't check it though

File: packages/vuetify/src/mixins/menuable.ts [hunk 2]
```diff
@@ -138,11 +164,11 @@ export default Vue.extend({
 
       return top
     },
-    hasActivator () {
+    hasActivator (): string | boolean | object {
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/mixins/menuable.ts]: `has` suggest a boolean, if it doesn't return boolean then it should

File: packages/vuetify/src/mixins/menuable.ts [hunk 3]
```diff
@@ -310,22 +336,22 @@ export default Vue.extend({
...
-        rect.left = parseInt(style.marginLeft)
-        rect.top = parseInt(style.marginTop)
+        rect.left = parseInt(style.marginLeft as string)
+        rect.top = parseInt(style.marginTop as string)
       }
 
       return rect
     },
-    sneakPeek (cb) {
+    sneakPeek (cb: Function) {
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/mixins/menuable.ts]: `cb: () => void`

File: packages/vuetify/src/mixins/menuable.ts [hunk 4]
```diff
@@ -310,22 +336,22 @@ export default Vue.extend({
...
       const rect = this.getRoundedBoundedClientRect(el)
 
       // Account for activator margin
       if (this.isAttached) {
-        const style = window.getComputedStyle(el)
+        const style = window.getComputedStyle(el) as CSSStyleDeclaration
 
-        rect.left = parseInt(style.marginLeft)
-        rect.top = parseInt(style.marginTop)
+        rect.left = parseInt(style.marginLeft as string)
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/mixins/menuable.ts]: `        rect.left = parseInt(style.marginLeft!)`

File: packages/vuetify/src/mixins/menuable.ts [hunk 5]
```diff
@@ -299,7 +325,7 @@ export default Vue.extend({
       return window.pageYOffset ||
         document.documentElement.scrollTop
     },
-    getRoundedBoundedClientRect (el) {
+    getRoundedBoundedClientRect (el: Element): Record<string, number> {
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/mixins/menuable.ts]: there should be specified interface for the returned object (top/left/bottom/right/width/height)

File: packages/vuetify/src/mixins/menuable.ts (hunk 2 above)
Comment (MajesticPotatoe) [reply packages/vuetify/src/mixins/menuable.ts]: trying to set this to boolean gives `Type 'string | boolean | object' is not assignable to type 'boolean'.
  Type 'string' is not assignable to type 'boolean'.` seems like activator can be type of string or object. not sure if thats right

Comment (jacekkarczmarczyk) [reply packages/vuetify/src/mixins/menuable.ts]: `return ... || !!this.activator || !!this.inputActivator`?

File: packages/vuetify/src/mixins/menuable.ts (hunk 5 above)
Comment (MajesticPotatoe) [reply packages/vuetify/src/mixins/menuable.ts]: seems like it maybe `ClientRect` ?

Comment (jacekkarczmarczyk) [reply packages/vuetify/src/mixins/menuable.ts]: Looks like it. Not sure why values are rounded here, but it's not in the scope of this issue

File: packages/vuetify/src/mixins/menuable.ts [hunk 6]
```diff
@@ -34,23 +56,28 @@ const dimensions = {
...
-
   props: {
     activator: {
       default: null,
-      validator: val => {
+      validator: (val: string | object) => {
         return ['string', 'object'].includes(typeof val)
       }
-    },
+    } as PropValidator<string | HTMLElement>,
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/mixins/menuable.ts]: that should be `string|HTMLElement|VNode` i guess

Comment (johnleider) [reply packages/vuetify/src/mixins/menuable.ts]: What would make this a VNode?

**Pull request merged**